### PR TITLE
tests: new test for snapshots with more than 1 user

### PR DESCRIPTION
**Pull request by sergiocazzolato**

### CODE REVIEW

**Comment by bboozzoo** [comment]
Shellcheck doesn't like some things about the test:
```
2018-11-09 01:47:12 Error executing google:ubuntu-18.04-64:tests/unit/spread-shellcheck : 
-----
+ testdir=/home/gopath/src/github.com/snapcore/snapd/tests/unit/spread-shellcheck
+ cd /home/gopath/src/github.com/snapcore/snapd
+ CAN_FAIL=/home/gopath/src/github.com/snapcore/snapd/tests/unit/spread-shellcheck/can-fail
+ ./spread-shellcheck spread.yaml tests
ERROR:root:shellcheck failed for file tests/main/snapshot-users/task.yaml in sections: execute; error log follows
ERROR:root:tests/main/snapshot-users/task.yaml: section 'execute':

In - line 109:
cat /home/test/snap/test-snapd-tools/current/canary/canary.txt | MATCH "content updated"
    ^-- SC2002: Useless cat. Consider 'cmd < file | ..' or 'cmd file | ..' instead.


In - line 110:
cat /home/test/snap/test-snapd-tools/current/canary.txt | MATCH "content updated"
    ^-- SC2002: Useless cat. Consider 'cmd < file | ..' or 'cmd file | ..' instead.

For more information:
  https://www.shellcheck.net/wiki/SC2002 -- Useless cat. Consider 'cmd < file | ....

ERROR:root:validation failed for the following non-whitelisted files:
 - tests/main/snapshot-users/task.yaml
-----
```

File: tests/main/snapshot-users/task.yaml [hunk 1]
```diff
@@ -0,0 +1,122 @@
+summary: Check that the basic snapshots functionality works for different users
+
+prepare: |
+    snap install test-snapd-tools
+
+execute: |
+    # use the snaps, so they create the dirs
+    test-snapd-tools.echo
+    su -p -c "/snap/bin/test-snapd-tools.echo" test
```
Comment (bboozzoo) [inline tests/main/snapshot-users/task.yaml]: Fedora and Arch do not have `/snap` by default:
```
#shellcheck source=tests/lib/dirs.sh
. "$TESTSLIB"/dirs.sh
su -p -c "$SNAP_MOUNT_DIR/bin/test-snapd-tools.echo" test
```

Comment (zyga) [reply tests/main/snapshot-users/task.yaml]: Can we just use `snap run test-snapd-tools.echo` instead?

Comment (zyga) [reply tests/main/snapshot-users/task.yaml]: This fails on:
```
/bin/bash: line 61: SNAP_MOUNT_DIR: unbound variable
```
We probably need to source dirs.sh

File: tests/main/snapshot-users/task.yaml [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    snap check-snapshot "$SET_ID_BOTH"
+    snap check-snapshot "$SET_ID_NONE"
+
+    # remove the canaries for both users
+    rm -f /root/snap/test-snapd-tools/{current,common}/canary.txt
+    rm -f /home/test/snap/test-snapd-tools/{current,common}/canary.txt
+
+    # restore the snapshot for the root user and check the files
+    snap restore "$SET_ID_ROOT" test-snapd-tools
+    test -e /root/snap/test-snapd-tools/current/canary.txt
```
Comment (mvo5) [inline tests/main/snapshot-users/task.yaml]: Should we check the content of the files here (and in the subsequent restores) as well?

File: tests/main/snapshot-users/task.yaml [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+    # restore the snapshot for both user and check the files
+    snap restore "$SET_ID_NONE" --users=test,root
+    test -e /root/snap/test-snapd-tools/current/canary.txt
+    test -e /root/snap/test-snapd-tools/common/canary.txt
+    test -e /home/test/snap/test-snapd-tools/current/canary.txt
+    test -e /home/test/snap/test-snapd-tools/common/canary.txt
+    test -e /home/test/snap/test-snapd-tools/current/canary/canary.txt
+
+    # check files content
```
Comment (mvo5) [inline tests/main/snapshot-users/task.yaml]: Should we run similar checks for the previous restores too?

File: tests/main/snapshot-users/task.yaml [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    MATCH "content updated" < /home/test/snap/test-snapd-tools/current/canary/canary.txt
+    MATCH "content updated" < /home/test/snap/test-snapd-tools/current/canary.txt
+
+    # check removal works
+    snap forget "$SET_ID_NONE"
+    snap saved --id="$SET_ID_NONE" | grep "No snapshots found"
```
Comment (mvo5) [inline tests/main/snapshot-users/task.yaml]: It looks like we want the same checks for the `snap forget` calls below. Maybe also a `[ $(snap saved | wc -l) = 0 ]` or something at the end?

Comment (sergiocazzolato) [reply tests/main/snapshot-users/task.yaml]: I implemented in that way to be consisten with the other snapshots tests, I'll see if there another better way to check that.

File: tests/main/snapshot-users/task.yaml (hunk 3 above)
Comment (sergiocazzolato) [reply tests/main/snapshot-users/task.yaml]: done

File: tests/main/snapshot-users/task.yaml (hunk 2 above)
Comment (sergiocazzolato) [reply tests/main/snapshot-users/task.yaml]: done

**Comment by sergiocazzolato** [comment]
@mvo5 , thanks for reviewing it, could you please take a look again?

File: tests/main/snapshot-users/task.yaml (hunk 1 above)
Comment (sergiocazzolato) [reply tests/main/snapshot-users/task.yaml]: done

**Comment by sergiocazzolato** [comment]
@chipaca, I think the main differences are:
. it is checking not just files but subdirectories are correctly snapshoted
. it is checking the restore for different users and both users and validating the content
. it is validating the option --users for the restore and save

The idea of this test is to complement the snapshot-cross-revno one, perhaps we also could just add some parts this one?

**Pull request merged**